### PR TITLE
[core] Fix layer impl cast in render layer implementations

### PR DESCRIPTION
**Pull request by pozdnyakov**
This PR puts `impl()` functions definitions into a nameless namespace
to provide internal linkage and to make sure that appropriate cast function
is invoked.

The problem was discovered by sanitize bot https://circleci.com/gh/mapbox/mapbox-gl-native/316959

### CODE REVIEW

**Comment by chloekraw** [comment]
@tmpsantos @pozdnyakov, thanks. Please add a changelog entry and request me for review on this changelog entry before merging.

**Comment by pozdnyakov** [comment]
> @tmpsantos @pozdnyakov, thanks. Please add a changelog entry and request me for review on this changelog entry before merging.

Done in the latest commit, please take a look.

File: platform/android/CHANGELOG.md [hunk 1]
```diff
@@ -10,6 +10,7 @@ Mapbox welcomes participation and contributions from everyone.  If you'd like to
  - Add fallback support to local ideograph font families [#15255](https://github.com/mapbox/mapbox-gl-native/pull/15255)
 
 ### Bug fixes
+ - Fixed possible crash caused by invoking of the wrong cast layer implementation function [#15398](https://github.com/mapbox/mapbox-gl-native/pull/15398).
```
Comment (chloekraw) [inline platform/android/CHANGELOG.md]: add "a" before "possible crash"

Comment (chloekraw) [reply platform/android/CHANGELOG.md]: since the crash is only "possible", the verb clause should be "that could be caused by"

Comment (chloekraw) [reply platform/android/CHANGELOG.md]: delete "of" --> "invoking the wrong"

Comment (chloekraw) [reply platform/android/CHANGELOG.md]: the last four words are better as: "layer casting implementation function."

File: platform/ios/CHANGELOG.md [hunk 2]
```diff
@@ -19,6 +19,7 @@ Mapbox welcomes participation and contributions from everyone. Please read [CONT
 
 ### Other changes
 
+* Fixed possible crash caused by invoking of the wrong cast layer implementation function. ([#15398](https://github.com/mapbox/mapbox-gl-native/pull/15398))
```
Comment (chloekraw) [inline platform/ios/CHANGELOG.md]: same comments as in android changelog

File: platform/android/CHANGELOG.md (hunk 1 above)
Comment (pozdnyakov) [reply platform/android/CHANGELOG.md]: layer implementation casting function - it casts `Layer::Impl` instances.

Comment (chloekraw) [reply platform/android/CHANGELOG.md]: ah yeah, that’s perfect, thanks!

Comment (pozdnyakov) [reply platform/android/CHANGELOG.md]: Thank you!

**Comment by chloekraw** [comment]
> this needs backporting, should be a straightforward cherry-pick

@pozdnyakov @tmpsantos could you clarify, which commit was this regression introduced in and which releases require backporting?

**Comment by pozdnyakov** [comment]
> @pozdnyakov @tmpsantos could you clarify, which commit was this regression introduced in and which releases require a backport?

It's https://github.com/mapbox/mapbox-gl-native/commit/4b21560cf59877125ea0bdae1a2546ab06f1efb2 . Also, the issue would affect only debug build configurations, where these casting functions do not get inlined.

**Comment by LukasPaczos** [comment]
> Also, the issue would affect only debug build configurations, where these casting functions do not get inlined.

On Android, we're shipping only release builds of GL-native in public artifacts so this wouldn't need patching if I understand correctly.

**Comment by pozdnyakov** [comment]
> On Android, we're shipping only release builds of GL-native in public artifacts so this wouldn't need patching if I understand correctly.

IMO still better to patch, in case someone is building SDK from source and with modified build flags.

**Comment by LukasPaczos** [comment]
This wouldn't require patch releases though, only CPs to release branches, right?

**Comment by LukasPaczos** [comment]
Capturing from a chat with @pozdnyakov, we tested `arm-v7`, `arm-v8` and `x86` builds and it looks like release builds that **we** publish are stable and do not crash. Therefore, we might not need to push additional patch releases, only CP the fix to release branches for developers that build from source.

That said, I'll let you @pozdnyakov @chloekraw to make the final call. If we'd like to be super-safe, we can schedule the patch releases as well.

/cc @tobrun

**Comment by pozdnyakov** [comment]
Here is the branch, reproducing the issue https://github.com/mapbox/mapbox-gl-native/tree/mikhail_reproduce_layer_impl_cast_problem

**Pull request merged**